File: filelist/src/main/scala/farjs/filelist/popups/MakeFolderController.scala
package farjs.filelist.popups

import farjs.filelist.FileListUiData
import farjs.filelist.history.{History, HistoryKind, HistoryProvider}
import scommons.react._
import scommons.react.hooks._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.scalajs.js

object MakeFolderController extends FunctionComponent[FileListUiData] {
  
  val mkDirsHistoryKind: HistoryKind = HistoryKind("farjs.mkdirs", 50)

  private[popups] var makeFolderPopup: ReactClass = MakeFolderPopup

  private var initialMultiple = false

  protected def render(compProps: Props): ReactElement = {
    val historyProvider = HistoryProvider.useHistoryProvider()
    val (multiple, setMultiple) = useState(initialMultiple)
    val props = compProps.plain

    props.data.toOption match {
      case Some(data) if props.showMkFolderPopup.getOrElse(false) =>
        <(makeFolderPopup)(^.plain := MakeFolderPopupProps(
          multiple = multiple,
          onOk = { (dir, multiple) =>
            val action = data.actions.createDir(
              dispatch = data.dispatch,
              parent = data.state.currDir.path,
              dir = dir,
              multiple = multiple
            )
            for {
              _ <- action.task.result.toFuture
              mkDirsHistory <- historyProvider.get(mkDirsHistoryKind).toFuture
              _ <- mkDirsHistory.save(History(dir, js.undefined)).toFuture
            } yield {
              setMultiple(multiple)
              initialMultiple = multiple

              props.onClose()
            }
            data.dispatch(action)
          },
          onCancel = props.onClose
        ))()
      case _ => null
    }
  }
}

File: filelist/popups/MakeFolderController.mjs
/**
 * @typedef {import("../FileListUi.mjs").FileListUiData} FileListUiData
 */
import React from "react";
import HistoryProvider from "@farjs/filelist/history/HistoryProvider.mjs";
import MakeFolderPopup from "./MakeFolderPopup.mjs";

const h = React.createElement;

let initialMultiple = false;

/**
 * @param {FileListUiData} props
 */
const MakeFolderController = (props) => {
  const { makeFolderPopup } = MakeFolderController;

  const historyProvider = HistoryProvider.useHistoryProvider();
  const data = props.data;

  if (data && props.showMkFolderPopup) {
    /** @type {(dir: string, multiple: boolean) => Promise<void>} */
    const saveHistory = async (dir, multiple) => {
      const mkDirsHistory = await historyProvider.get(
        MakeFolderPopup.mkDirsHistoryKind
      );
      await mkDirsHistory.save({ item: dir });

      initialMultiple = multiple;
      props.onClose();
    };

    return h(makeFolderPopup, {
      multiple: initialMultiple,
      onOk: (dir, multiple) => {
        const action = data.actions.createDir(
          data.dispatch,
          data.state.currDir.path,
          dir,
          multiple
        );
        const a = {
          ...action,
          task: {
            ...action.task,
            result: action.task.result.then(async (res) => {
              await saveHistory(dir, multiple);
              return res;
            }),
          },
        };
        data.dispatch(a);
      },
      onCancel: props.onClose,
    });
  }

  return null;
};

MakeFolderController.displayName = "MakeFolderController";
MakeFolderController.makeFolderPopup = MakeFolderPopup;

export default MakeFolderController;
